### PR TITLE
Created a broker-plugin, with the customization's to the broker for EnMasse

### PR DESCRIPTION
**Pull request by vbusch**
Separated EnMasse configurations into a broker-plugin. 

Signed-off-by: Vanessa <vbusch@localhost.localdomain>

### CODE REVIEW

File: broker-plugin/plugin/src/main/sh/launch-broker.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    cp -n $BROKER_CUSTOM_DIR/certs/* $instanceDir/etc/
+    echo "cp -n $BROKER_CUSTOM_DIR/certs/* $instanceDir/etc/"
+}
+
+
+# This needs to be at the toplevel outside any functions
+# For the standard address space, the shutdown hooks need time to run before broker is shut down
+if [ "$ADDRESS_SPACE_TYPE" != "brokered" ]; then
+    trap "" TERM INT
+fi
```
Comment (lulf) [inline broker-plugin/plugin/src/main/sh/launch-broker.sh]: The trap can be removed now that the shutdown hook is disabled

File: templates/address-space-controller/020-ConfigMap-address-space-definitions.yaml [hunk 2]
```diff
@@ -1415,3 +1454,11 @@ data:
...
       name: BROKER_ADDRESS_FULL_POLICY
       value: FAIL
+    - description: The name of the broker
+      displayName: AMQ Name
+      name: AMQ_NAME
+      value: broker
+    - description: Home directory to for the broker instance
+      displayName: Home
+      name: HOME
+      value: /home/artemis
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-address-space-definitions.yaml]: I'm not sure it makes sense to templatize these values. I think instead we should just set them in the pod container env.

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml [hunk 3]
```diff
@@ -115,15 +109,26 @@ data:
...
+                value: ${AMQ_NAME}                
               image: ${ARTEMIS_IMAGE}
               imagePullPolicy: ${IMAGE_PULL_POLICY}
+              command:
+              - /opt/apache-artemis/custom/bin/launch-broker.sh
+              lifecycle:
+                preStop:
+                  exec:
+                    command:
+                    - $ARTEMIS_HOME/custom/shutdown-hook/shutdown-hook.sh
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: Remove the lifecycle hook as it has been removed from master.

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml [hunk 4]
```diff
@@ -115,15 +109,26 @@ data:
               - name: AUTHENTICATION_SERVICE_SASL_INIT_HOST
                 value: ${AUTHENTICATION_SERVICE_SASL_INIT_HOST}
               - name: AUTHENTICATION_SERVICE_OAUTH_URL
-                value: ${AUTHENTICATION_SERVICE_OAUTH_URL}
+                value: ${AUTHENTICATION_SERVICE_OAUTH_URL}                
+              - name: HOME
+                value: ${HOME}
+              - name: AMQ_NAME
+                value: ${AMQ_NAME}
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: hardocde values

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml [hunk 5]
```diff
@@ -229,6 +281,14 @@ data:
...
       name: GLOBAL_MAX_SIZE
       value: "-1"
+    - description: The name of the broker
+      displayName: AMQ Name
+      name: AMQ_NAME
+      value: broker
+    - description: Home directory to for the broker instance
+      displayName: Home
+      name: HOME
+      value: /home/artemis
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: Remove template parameters and hardcode value in pod container env

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml [hunk 6]
```diff
@@ -340,15 +394,26 @@ data:
...
+                value: ${AUTHENTICATION_SERVICE_OAUTH_URL}                
+              - name: HOME
+                value: ${HOME}
+              - name: AMQ_NAME
+                value: ${AMQ_NAME}                
               image: ${ARTEMIS_IMAGE}
               imagePullPolicy: ${IMAGE_PULL_POLICY}
+              command:
+              - /opt/apache-artemis/custom/bin/launch-broker.sh
+              lifecycle:
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: remove lifecycle hook as on master

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml [hunk 7]
```diff
@@ -483,3 +595,11 @@ data:
...
       name: GLOBAL_MAX_SIZE
       value: "-1"
+    - description: The name of the broker
+      displayName: AMQ Name
+      name: AMQ_NAME
+      value: broker
+    - description: Home directory to for the broker instance
+      displayName: Home
+      name: HOME
+      value: /home/artemis
```
Comment (lulf) [inline templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: Remove template parameters and specify values in pod container env instead.

File: artemis/Dockerfile [hunk 8]
```diff
@@ -1,14 +1,20 @@
...
+RUN yum -y install which libaio python hostname iputils && yum clean all -y && mkdir -p /var/run/artemis/
 
 ARG version
 ARG commit
-ENV ARTEMIS_HOME=/opt/apache-artemis-2.6.3 PATH=$ARTEMIS_HOME/bin:$PATH VERSION=${version} COMMIT=${commit} PLUGIN_HOME=/opt/broker-plugin
+
+ENV ARTEMIS_HOME=/opt/apache-artemis HOME=/home/artemis PATH=$ARTEMIS_HOME/bin:$PATH VERSION=${version} COMMIT=${commit}
 
 ADD ./build/apache-artemis-bin.tar.gz /opt
+RUN mv /opt/apache-artemis-2.6.3 $ARTEMIS_HOME
```
[the remark this answers is not in the document]
Comment (vbusch) [reply artemis/Dockerfile]: 'apache-artemis-2.6.3' is the directory name inside of the archive. If I extract it to '/opt/apache-artemis', I will end up with '/opt/apache-artemis/apache-artemis-2.6.3'.  I could change $ARTEMIS_HOME to include the version.

File: broker-plugin/plugin/src/main/sh/init-broker.sh [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+    export HAWTIO_ROLE=manage
+}
+
+function pre_configuration() {
+    local instanceDir="${HOME}/${AMQ_NAME}"
+    
+    echo "export AMQ_USER=admin" >> $BROKER_CUSTOM/bin/env.sh
+    echo "export AMQ_PASSWORD=admin" >> $BROKER_CUSTOM/bin/env.sh
+    echo "export AMQ_REQUIRE_LOGIN=false" >> $BROKER_CUSTOM/bin/env.sh
+    echo "export AMQ_TRANSPORTS=amqp" >> $BROKER_CUSTOM/bin/env.sh
```
[the remark this answers is not in the document]
Comment (vbusch) [reply broker-plugin/plugin/src/main/sh/init-broker.sh]: Yes, the broker.xml from the configure_brokered script will add the support for the other protocols.

File: broker-plugin/plugin/src/main/sh/launch-broker.sh (hunk 1 above)
Comment (vbusch) [reply broker-plugin/plugin/src/main/sh/launch-broker.sh]: now removed

File: templates/address-space-controller/020-ConfigMap-address-space-definitions.yaml (hunk 2 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-address-space-definitions.yaml]: now removed

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml (hunk 3 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: now removed

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml (hunk 4 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: now removed

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml (hunk 5 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: now removed

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml (hunk 6 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: now removed

File: templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml (hunk 7 above)
Comment (vbusch) [reply templates/address-space-controller/020-ConfigMap-standard-broker-definitions.yaml]: now removed

**Pull request merged**